codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3GetBucketLocation.java

```diff
@@ -0,0 +1,209 @@
+                                    .equals(getBucketLocationOpID);
+                        })
+                        .registerMiddleware(MiddlewareRegistrar.builder()
+                                .resolvedFunction(
+                                        SymbolUtils.createValueSymbolBuilder(swapDeserializerFuncName).build())
+                                .build())
+                        .build()
+        );
+    }
+
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoDelegator goDelegator
+    ) {
+        ShapeId serviceId = settings.getService();
+        ServiceShape service = model.expectShape(serviceId, ServiceShape.class);
+        if (!isS3Service(model, service)) {
+            return;
+        }
+
+        for (ShapeId operationId : service.getAllOperations()) {
+            if (!(operationId.getName().equals(getBucketLocationOpID))) {
+                continue;
+            }
+
+            OperationShape operation = model.expectShape(operationId, OperationShape.class);
+            goDelegator.useShapeWriter(operation, writer -> {
+                writeCustomDeserializer(writer, model, symbolProvider, operation);
+                writeDeserializerSwapFunction(writer, operation);
+            });
+        }
+
+    }
+
+    /**
+     * writes helper function to swap deserialization middleware with the generated
+     * custom deserializer middleware.
+     *
+     * @param writer    is the go writer used
+     * @param operation is the operation for which swap function is written.
+     */
+    private void writeDeserializerSwapFunction(
+            GoWriter writer,
+            OperationShape operation
+    ) {
+        writer.writeDocs("Helper to swap in a custom deserializer");
+        writer.openBlock("func $L(stack *middleware.Stack) error{", "}",
+                swapDeserializerFuncName, () -> {
+                    writer.write("_, err := stack.Deserialize.Swap($S, &$L{})",
+                            ProtocolUtils.OPERATION_DESERIALIZER_MIDDLEWARE_ID.getString(),
+                            getDeserializeMiddlewareName(operation)
+                    );
+                    writer.write("if err != nil { return err }");
+                    writer.write("return nil");
+                });
+    }
+
+    /**
+     * writes a custom deserializer middleware for the provided operation.
+     *
+     * @param goWriter       is the go writer used.
+     * @param model          is the generation model.
+     * @param symbolProvider is the symbol provider.
+     * @param operation      is the operation shape for which custom deserializer is written.
+     */
+    private void writeCustomDeserializer(
+            GoWriter goWriter,
+            Model model,
+            SymbolProvider symbolProvider,
+            OperationShape operation
+    ) {
```

Comment:
is there anyway this could reuse the existing deserializer generator? This duplicates a lot of behavior from the generic deserializer. I think we need to investigate how the deserializers can be extended and augmented with custom behavior. Because the only thing that changes here is how the response body document is deserialized.

Ideally codegen generates everything the same, but the shape GetBucketLocationOutput deserializer func has custom behavior instead of duplicating the operation deserializer as well.

Reply:
Keeping this deserializer and regular deserializer in parity adds an extra burden that could lead to unexpected divergences unless they share the same API.

Reply:
We reuse the deserializers for member shapes. The top level deserializer function is generated via smithy-go. This custom deserializer generates a custom top level deserializer function, and uses the existing member deserializer function. 

Best case would be able to override the deserializer function method, and be able to override it for a specific operation shape. But seems unnecessary for a handful of operations that do not follow the protocol pattern.

Reply:
Yeah this change is an artifact of the (de)serializer generators not being too rigid in its capabilities. This is something we should be looking at how we can improve in the long term. Creating duplicate code is definitely not a long term solution, and will seriously hurt maintainability in the future.